PageObjects/ShopPage.cs

```diff
@@ -1,43 +1,139 @@
+            var filter = page.Locator(filterCheckboxLocator);
+
+            await filter.CheckAsync();
+            await Assertions.Expect(filter).ToBeCheckedAsync();
+        }
 
-        #region Page
-        public async Task GoToTestPageURL()
+        [Obsolete]
+        public async Task VerifyFilteredProducts(string filterValue)
         {
-            await Page.GotoAsync(testPageUrl);
+            await page.WaitForNavigationAsync();
+            var productTitleLocator = "//*[@class[starts-with(., 'list-product-title')]]";
+            var allProducts = await page.Locator(productTitleLocator).AllInnerTextsAsync();
+            var productsList = allProducts.ToList();
+            Assert.That(productsList.Count, Is.GreaterThan(0), $"Products by locator {productTitleLocator} not found");
+
+            bool isAllContainFilterValue = productsList.All(product => product.ToLower().Contains(filterValue.ToLower()));
+            Assert.That(isAllContainFilterValue, Is.True, $"Not all Products contains the text {filterValue}");
         }
-        public async Task WaitForUrlLoading()
+
+        public async Task AddSpecifiedProductToCart(string addProduct)
         {
-            await Page.WaitForURLAsync(testPageUrl);
+            var productHolderLocator = "//*[contains(@class, 'card z-depth-1 hoverable')]";
+            var addToCartBtnLocator = "//*[@class[starts-with(., 'add-product-to-cart')]]";
+            var addPopup = page.Locator("//*[@id='cart-modal']");
+
+            // Get all product holders on the page
+            var productCards = await page.QuerySelectorAllAsync(productHolderLocator);
+            Assert.That(productCards, Is.Not.Empty, "Products not found on the page");
+
+            // Iterate through each product holder to find the {addProduct}
+            bool isProductFound = false; // To track if product is found
+            foreach (var productCard in productCards)
+            {
+                var productName = await productCard.InnerTextAsync();
+                if (productName.ToLower().Contains(addProduct.ToLower()))
+                {
+                    isProductFound = true;
+                    var addToCartButton = await productCard.QuerySelectorAsync(addToCartBtnLocator);
+                    Assert.That(addToCartButton, Is.Not.Null, $"'Add to cart button' not found for product {addProduct}");
+
+                    //Add specified product to the cart
+                    if (addToCartButton != null)
+                    {
+                        await addToCartButton.ClickAsync();
+                    }
+                    await Assertions.Expect(addPopup).ToContainTextAsync($"{addProduct}");
+                    await Assertions.Expect(addPopup).ToContainTextAsync($"Товар додано у кошик");
+
+                    return;
+                }
+            }
+            Assert.That(isProductFound, Is.True, $"Product '{addProduct}' is not found on the page");
```

Comment:
this entire implementation should be just 2 lines of code:
1) create locator for add to cart icon based on "string productDescription" input
2) click on this locator
or "Filtering Locators" approach from https://playwright.dev/dotnet/docs/locators

Verification that product is added - is better to have as a separate method